File: backend/app/application/ports/services.py
from typing import Protocol
from uuid import UUID

from app.application.contracts.audio import AudioSample
from app.domain.analysis import Analysis


class AnalysisQuotaExhausted(Exception):
    """Raised by the grammar analyzer when quota is reached, and by the quote enforcer"""


class PasswordHasher(Protocol):
    def hash(self, password: str) -> str: ...
    def verify(self, password: str, password_hash: str) -> bool: ...


class TokenService(Protocol):
    def issue(self, user_id: UUID) -> str: ...
    def verify(self, token: str) -> UUID | None: ...


class GrammarAnalyzer(Protocol):
    async def analyze(self, audio: AudioSample) -> tuple[str, Analysis]: ...


class AnalysisQuotaEnforcer(Protocol):
    async def try_consume(self, user_id: UUID) -> bool: ...

File: src/application/ports/services.ts
import type { AudioSample } from "../contracts/audio";
import type { StagedAudioReference } from "../contracts/staged-audio";
import type { Analysis } from "../../domain/analysis";
import type { UserId } from "../../domain/user";

export interface PasswordHasher {
  hash(password: string): Promise<string>;
  verify(password: string, passwordHash: string): Promise<boolean>;
}

export interface TokenService {
  issue(userId: UserId): Promise<string>;
  verify(token: string): Promise<UserId | null>;
}
export interface SpeechAnalysisResult {
  readonly analysis: Analysis;
  readonly transcript: string;
}
export interface SpeechAnalyzer {
  analyze(audio: AudioSample): Promise<SpeechAnalysisResult>;
}

export interface AnalysisQuota {
  tryConsume(userId: UserId): Promise<boolean>;
}

/**
 * Resolves temporary audio without exposing a storage vendor or URL to the
 * application. Implementations must enforce that the reference belongs to the
 * authenticated user.
 */
export interface StagedAudioStore {
  resolve(
    userId: UserId,
    reference: StagedAudioReference,
  ): Promise<AudioSample>;
  delete(userId: UserId, reference: StagedAudioReference): Promise<void>;
}
